src/client.js

```diff
@@ -0,0 +1,97 @@
+"use strict";
+var W3CWebSocket = require('websocket').w3cwebsocket;
+class Client {
+  constructor(url, protocol) {
+    this.url = url;
+    this.protocol = protocol;
+    this.client = new W3CWebSocket(url, protocol);
+    this.subscriptions = {}; // id: handler
+    this.max_id = 0;
+
+    this.client.onopen = () => {
+      console.log("Websocket client connected.");
+    };
+
+    this.client.onmessage = (message) => {
+      let message_data = JSON.parse(message.data);
+      if (message_data.type === 'subscription_fail') {
+        const del_id = message_data.id;
+        delete this.subscriptions[del_id];
+      } else if (message_data.type === 'subscription_data') {
+        if (message_data.data) {
+          const sub_id = message_data.id;
+          if (this.subscriptions[sub_id]) {
+            this.subscriptions[sub_id](null, message_data.data); // pass data into data handler
+          }
```

Comment:
If this.subscriptions[sub_id] doesn't exist, maybe we should send a 'subscription_end' to the server.